numpy/core/src/multiarray/dragon4.c

```diff
@@ -640,8 +699,8 @@ static void
 BigInt_Pow10(BigInt *result, npy_uint32 exponent)
 {
     /* create two temporary values to reduce large integer copy operations */
-    BigInt temp1;
-    BigInt temp2;
+    static BigInt temp1;
+    static BigInt temp2;
```

Comment:
What gain is there to making these static?

Reply:
Maybe this is a bad idea... but I was worried about using up too much stack space. Each BigInt is 4k.

Reply:
(I didn't want to use stack beause these are big variables, and I didn't want to use the heap cause it's slower. This way is uses the "data" memory.)

Reply:
Do these actually need to be big, or are some of these bigints only int128s? Could you use `alloca` to allocate bigints only of the size needed on the stack?

Reply:
The BigInts have to be able to grow as the computation progresses. Most of the time the BigInt needs much less than 4k, but sometimes (very hard to know beforehand) it needs a lot more. I think in all cases they can potentially be more than int128.

By the way, there may be some optimizations to make here. Ryan Jucket optimized his BigInts for speed, but his were only 140 bytes and he saved them on the stack, because he only did up to 64 bit floats. But for float128 I need much bigger ints. 

For float64, the max exponent (base 2) is 1023, so we need BigInts up to 1023 bits, plus 53 for the mantissa. (140 bytes = 1120 bits). For float128, the max exponent (base 2) is 16383, so we need 16383 bits plus 64 for the mantissa, or about 2kb. The 4k I used is a bit overkill.

Reply:
Is it thread safe? Putting things on the stack has that advantage.

Reply:
No, it isn't. I'm still hesitant about the stack.. but let me try mallocing them on the heap, and see if there's a performance hit.

From some quick performance tests I did a while ago, it seemed possible there is some python/numpy overhead to printing that might dwarf the cost of Dragon4 itself, in which case the speed of the BigInt allocation might be negligible.

Reply:
I did some measurement:

For this PR currently, using "data" memory, I get that float64 printing with Dragon4 is about 5% slower that `printf("%.18g"`, when printing numbers generated by `np.random.rand`.

When I switch to malloc'ing on the heap, there is a pretty big slowdown, so that Dragon4 becomes about 300% slower than printf (i.e., a factor of 3). I tried to optimize by malloc'in all the needed BigInts at once to get a single malloc.

Is it a big deal if the repr is not re-entrant? We don't release the GIL for printing anywhere, so I would favor keeping the `static`.

We can also consider leaving the BigInts in the stack space: The total space needed by a single call to Dragon4 is about 24kb (6 BigInts). Typical stacks are 8mb (linux) and 1mb (windows). The goal of putting this into stack instead of data is to make Dragon4 re-entrant, but with such a large stack increase there is a real danger of using up the stack when re-entering. So even if we allocate in stack, we should recommend *not* to re-enter. So might as well allocate in "data".